euphonic/spectra.py

```diff
@@ -229,62 +242,69 @@ class Spectrum1D(Spectrum):
-        size (n_x_data + 1,))
-    y_data : (n_x_data,) float Quantity
-        The plot data in y
-    x_tick_labels : list (int, string) tuples or None
+    x_data
+        Shape (n_x_data,) or (n_x_data + 1,) float Quantity. The x_data
+        points (if size == (n_x_data,)) or x_data bin edges (if size
+        == (n_x_data + 1,))
+    y_data
+        Shape (n_x_data,) float Quantity. The plot data in y
+    x_tick_labels : Sequence[Tuple[int, str]] or None
         Special tick labels e.g. for high-symmetry points. The int
         refers to the index in x_data the label should be applied to
+    metadata : Dict[str, Any] or None
+        Contains metadata about the spectrum. Any keys/values are
+        allowed, but values must be JSON serialisable to write to a
+        json file. There are some functional keys:
+          - 'label' : str. This is used label lines on a 1D plot
     """
-    def __init__(self, x_data, y_data, x_tick_labels=None):
+    def __init__(self, x_data: Quantity, y_data: Quantity,
+                 x_tick_labels: Optional[Sequence[Tuple[int, str]]] = None,
+                 metadata: Optional[Dict[str, Any]] = None) -> None:
         """
         Parameters
         ----------
-        x_data : (n_x_data,) or (n_x_data + 1,) float Quantity
-            The x_data points (if size (n_x_data,)) or x_data bin edges
-            (if size (n_x_data + 1,))
-        y_data : (n_x_data,) float Quantity
-            The plot data in y
-        x_tick_labels : list (int, string) tuples or None
+        x_data
+            Shape (n_x_data,) or (n_x_data + 1,) float Quantity. The
+            x_data points (if size == (n_x_data,)) or x_data bin edges
+            (if size == (n_x_data + 1,))
+        y_data
+            Shape (n_x_data,) float Quantity. The plot data in y
+        x_tick_labels
             Special tick labels e.g. for high-symmetry points. The int
             refers to the index in x_data the label should be applied to
+        metadata
+            Contains metadata about the spectrum. Any keys/values are
+            allowed, but values must be JSON serialisable to write to a
+            json file. There are some functional keys:
+              - 'label' : str. This is used label lines on a 1D plot
         """
         _check_constructor_inputs(
-            [y_data, x_tick_labels],
-            [Quantity, [list, type(None)]],
-            [(-1,), ()],
-            ['y_data', 'x_tick_labels'])
+            [y_data, x_tick_labels, metadata],
+            [Quantity, [list, type(None)], [dict, type(None)]],
+            [(-1,), (), ()],
+            ['y_data', 'x_tick_labels', 'metadata'])
         ny = len(y_data)
         _check_constructor_inputs(
             [x_data], [Quantity],
             [[(ny,), (ny+1,)]], ['x_data'])
         self._set_data(x_data, 'x')
         self._set_data(y_data, 'y')
         self.x_tick_labels = x_tick_labels
+        self.metadata = metadata
```

Comment:
As suggested in plotting:
```python
self.metadata = {} if metadata is None else metadata
```
- guarantee that metadata is always a dict and can be queried with `in` or `get`; saves on lots of type-checking elsewhere
- we still use `None` as the default empty argument to avoid Python pass-by-reference nasties

Maybe we should think think about using `dict.copy()` to avoid mutation as well. Which behaviour is more/less surprising?

Reply:
That would make the overall logic simpler, the only slight issue is that `"metadata": {}` would have to be added to every specrum JSON output in the tests. Unless I add something to avoid writing an empty metadata attribute? There is already logic to omit `None` values when converting to dicts/json. Which would be preferable, adding metadata to every .json or adding logic to omit it if it's empty?

The copy behaviour is a little inconsistent currently - if a `Quantity` is passed it gets copied during unit conversion, but if an array is passed it is just set as the attribute and no copy is made. We could set the behaviour to always copy but for things like eigenvectors this could potentially be quite expensive... I guess for the case of metadata it makes sense to copy for now as it should be quite small, but we should probably decide on something consistent or at least document it somewhere...

Reply:
I'll add some logic to omit empty metadata from outputs for now so that the tests keep passing, and then update the test data later if needed.

Reply:
Omitting empty metadata from JSON seems fine to me. Think of it as part of **kwargs for `__init__`, where missing keys will be replaced by the default (None) during instantiation.